Refactor pkg/kubelet/kubelet.go: syncPod().

yifan-gu opened the pull request:
Makes the syncPod() takes only the containers that belongs to the pod.

Commit:
Refactor pkg/kubelet/kubelet.go: syncPod().

Makes the syncPod() takes only the containers that belongs to the pod.

yifan-gu commented:
Ideally, I think it could be better for `syncPod()` to take only two pod level parameters. (e.g. `syncPod(pod *api.BoundPod, runningPod *api.Pod)`)

This can benefits us for a clean logic in syncPod, also it can makes the way easier for integrating `rocket` or other container runtime later.

Feedback highly appreciated! 
Thank you!

yifan-gu commented:
Next plan is to clean up the syncPod() itself.
I have currently working version of the new syncPod(), which can start/stop/update rocket containers in the pod. ()

I tried to reorganize the syncPod() in that branch (https://github.com/yifan-gu/kubernetes/blob/container_runtime/pkg/kubelet/kubelet.go#L1010), and also scratched a simple interface for container runtime here(https://github.com/yifan-gu/kubernetes/blob/container_runtime/pkg/kubelet/container.go)

erictune commented:
@pmorie this might intersect with the tmpfs changes you are making?  not sure.

@yujuhong this might intersect with what you are doing.

pmorie commented:
@erictune I changed my WIP for #4625 per @thockin's request to make the kubelet have only `getTmpfsPodVolumeDir`.  I'm not sure that we will need to create that automatically for every pod, but if we want to, it will touch my PR, yes.

vmarmol commented on pkg/kubelet/dockertools/docker.go:
```diff
@@ -413,17 +413,22 @@ func (c DockerContainers) FindPodContainer(podFullName string, uid types.UID, co
+}
+
+// FindContainersByPod returns the containers that belong to the pod.
+func (c DockerContainers) FindContainersByPod(podUID types.UID, podFullName string) DockerContainers {
```
Only use UID, it is guaranteed to be unique in time and space. The podFullName is a vestige of an older era.

yifan-gu replied:
@vmarmol Good news to hear.

vmarmol commented on pkg/kubelet/dockertools/docker.go:
```diff
@@ -413,17 +413,22 @@ func (c DockerContainers) FindPodContainer(podFullName string, uid types.UID, co
-		if dockerManifestID == podFullName {
-			containers[dockerContainerName] = dockerContainer
+		dockerManifestID, uuid, _, _ := ParseDockerName(dockerContainer.Names[0])
+		if uuid == podUID && dockerManifestID == podFullName {
```
No need to check podFullName per comment above.

vmarmol commented on pkg/kubelet/kubelet_test.go:
```diff
@@ -874,11 +874,6 @@ func TestSyncPodDeletesDuplicate(t *testing.T) {
 			Names: []string{"/k8s_foo_bar.new.test_12345678_3333"},
 			ID:    "4567",
 		},
-		"2304": &docker.APIContainers{
```
Why removed from the test?

vmarmol replied:
Aaaaah right because this would no longer exist now. Got it. Would be nice if we have this test elsewhere, but that may be too hard for now.

yifan-gu replied:
@vmarmol Added a test for `FindContainersByPod`

vmarmol replied:
Thanks :D

vmarmol commented on pkg/kubelet/kubelet.go:
```diff
@@ -998,11 +998,24 @@ func (kl *Kubelet) killContainersInPod(pod *api.BoundPod, dockerContainers docke
+	return nil
+}
+
 func (kl *Kubelet) syncPod(pod *api.BoundPod, dockerContainers dockertools.DockerContainers) error {
```
nit: Maybe to go alongside this nice refactoring, rename dockerContainers to containersInPod?

vmarmol commented on pkg/kubelet/kubelet.go:
```diff
@@ -1045,7 +1051,6 @@ func (kl *Kubelet) syncPod(pod *api.BoundPod, dockerContainers dockertools.Docke
 			}
 		}
 	}
-	containersToKeep[podInfraContainerID] = empty{}
```
I think we're missing removing the POD infra container. We do bellow in the case of a change in the container, but if there is no change we don't seem to be doing it.

yifan-gu replied:
@vmarmol I think the `containersInPod.RemoveContainerWithID(podInfraContainerID)` bellow has taken care of this. That one is outside the supercalifragilisticexpialidocious for-loop. Even we don't change the container, we will hit that sentence.

After this, I will try to work on the prober inside the for-loop next to make it cleaner. :)

vmarmol replied:
Ah yes I see it now. Could we move it up to here instead? It seems to follow better here.

yifan-gu commented:
@vmarmol Thanks for the comments, I've addressed them, PTAL.

vmarmol commented:
This looks good to me @yifan-gu. Thanks for the simplificiation! Have you had a chance to run the e2e tests?

I only had one nit which would be nice to fix before we merge if you have the time :)

yifan-gu commented:
@vmarmol I did it. 
Also fixed the logic after `if podChanged` and changed `killContainerByID` to take only `ID`, (btw strangled a potential bug which passes empty string as `name` to `killContainerByID`)

dchen1107 commented:
I didn't look at PR in detail, but agreed with the goals posted here by @yifan-gu 

Only one comment related to podFullName gets my attention. Yes, podFullName was introduced initially before we had UID / UUID, and we should eventually get rid of them completely. But please note that there are containerVM user, which doesn't have UID assigned to their pods. Please make sure we won't break them here. We could use UID to query, if failed, retry with podFullName for now.

vmarmol commented:
Looks like we need a rebase. I also ran the e2e tests vs your PR before your last changes and 3 of them failed. Worth a look:

```
Summarizing 3 Failures:

[Fail] Pods [It] should be restarted with a /healthz http liveness probe 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/pods.go:52

[Fail] Pods [It] should be restarted with a docker exec "cat /tmp/health" liveness probe 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/pods.go:52

[Fail] kubectl [It] should scale a replication controller 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/kubectl.go:187

Ran 25 of 25 Specs in 2491.552 seconds
FAIL! -- 22 Passed | 3 Failed | 0 Pending | 0 Skipped
```

yujuhong commented:
FYI, my PR #5103 modifies GetPodFullName, but the PR doesn't get rid of it.

By the way, the kubernetes-e2e-gce dashboard shows some common failures such as "Pods should be restarted with a /healthz http liveness probe".

yifan-gu commented:
@vmarmol I run e2e on head, seems to have the same failures as @yujuhong said.
Revert `FindContainersByPod` to support empty pod uid as mentioned by @dchen1107

vmarmol commented:
Thanks! I just re-ran the e2e tests from HEAD and they're passing. I rebased your PR and am re-running them again. Will report back :)

Thanks for the heads up @yujuhong, I now see Jenkins is reporting problems but I run the tests and they seem fine so I'm confused :-\

vmarmol commented:
Happy to report that the e2e tests are clean :) thanks again @yifan-gu. LGTM and merging.

yifan-gu commented:
@vmarmol You're welcome:) Going to clean up the probing function next
